cpp/oneapi/dal/table/detail/table_iface.hpp

```diff
@@ -0,0 +1,157 @@
+};
+
+class homogen_table_iface : public table_iface {
+public:
+    virtual const void* get_data() const = 0;
+};
+
+class table_builder_iface {
+public:
+    virtual ~table_builder_iface() = default;
+    virtual table_iface* build() = 0;
+    virtual pull_rows_iface* get_pull_rows_iface() = 0;
+    virtual pull_column_iface* get_pull_column_iface() = 0;
+    virtual push_rows_iface* get_push_rows_iface() = 0;
+    virtual push_column_iface* get_push_column_iface() = 0;
+};
+
+class homogen_table_builder_iface : public table_builder_iface {
+public:
+    virtual homogen_table_iface* build_homogen() = 0;
+
+    virtual void set_data_type(data_type dt) = 0;
+    virtual void set_layout(data_layout layout) = 0;
+    virtual void set_feature_type(feature_type ft) = 0;
+
+    virtual void move(homogen_table_iface& t) = 0;
+    virtual void reset(const array<byte_t>& data,
+                       std::int64_t row_count,
+                       std::int64_t column_count) = 0;
+
+    virtual void allocate(std::int64_t row_count, //
+                          std::int64_t column_count) = 0;
+
+    virtual void copy_data(const void* data, //
+                           std::int64_t row_count,
+                           std::int64_t column_count) = 0;
+
+#ifdef ONEDAL_DATA_PARALLEL
+    virtual void allocate(const data_parallel_policy& policy,
+                          std::int64_t row_count,
+                          std::int64_t column_count,
+                          sycl::usm::alloc alloc) = 0;
+
+    virtual void copy_data(const data_parallel_policy& policy,
+                           const void* data,
+                           std::int64_t row_count,
+                           std::int64_t column_count) = 0;
+#endif
+};
+
+template <typename Iface, typename Derived>
+class table_template : public Iface,
+                       public pull_rows_template<Derived>,
+                       public pull_column_template<Derived> {
+public:
+    pull_rows_iface* get_pull_rows_iface() override {
+        return this;
+    }
+
+    pull_column_iface* get_pull_column_iface() override {
+        return this;
+    }
+};
+
+template <typename Iface, typename Derived>
+class table_builder_template : public Iface,
+                               public pull_rows_template<Derived>,
+                               public pull_column_template<Derived>,
+                               public push_rows_template<Derived>,
+                               public push_column_template<Derived> {
+public:
+    pull_rows_iface* get_pull_rows_iface() override {
+        return this;
+    }
+
+    pull_column_iface* get_pull_column_iface() override {
+        return this;
+    }
+
+    push_rows_iface* get_push_rows_iface() override {
+        return this;
+    }
+
+    push_column_iface* get_push_column_iface() override {
+        return this;
+    }
+};
+
+template <typename Derived>
+using dense_table_template = table_template<table_iface, Derived>;
```

Comment:
Why it is named dense? Per my understanding, even sparse tables can provide access to rows/columns in a dense way

Reply:
I'll remove aliases for `table_template` and will use it directly